ddtrace/contrib/anthropic/utils.py

```diff
@@ -0,0 +1,284 @@
+    elif _is_generator(resp):
+        return TracedOpenAIStream(resp, integration, span, args, kwargs)
+    elif _is_stream_manager(resp):
+        return TracedOpenAIStreamManager(resp, integration, span, args, kwargs)
+
+
+def _process_finished_stream(integration, span, args, kwargs, streamed_chunks, message_accumulated=False):
+    messages = None
+    messages = kwargs.get("messages", None)
+    chat_messages = get_argument_value(args, kwargs, 0, "messages")
+    try:
+        if message_accumulated:
+            messages = _construct_accumulated_messages(streamed_chunks)
+        else:
+            messages = _construct_unaccumulated_messages(streamed_chunks)
+        if integration.is_pc_sampled_span(span):
+            _tag_streamed_chat_completion_response(integration, span, messages)
+        if integration.is_pc_sampled_llmobs(span):
+            integration.llmobs_set_tags(
+                span,
+                input_messages=chat_messages,
+                formatted_response=messages,
+            )
+    except Exception:
+        log.warning("Error processing streamed completion/chat response.", exc_info=True)
+
+
+class BaseTracedOpenAIStream(wrapt.ObjectProxy):
+    def __init__(self, wrapped, integration, span, args, kwargs, is_message_stream=False):
+        super().__init__(wrapped)
+        n = kwargs.get("n", 1) or 1
+        self._dd_span = span
+        self._streamed_chunks = [[] for _ in range(n)]
+        self._dd_integration = integration
+        self._kwargs = kwargs
+        self._args = args
+        self._is_message_stream = (
+            is_message_stream  # message stream helper will have accumulated the message for us already
+        )
+
+
+class TracedOpenAIStream(BaseTracedOpenAIStream):
+    def __enter__(self):
+        self.__wrapped__.__enter__()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.__wrapped__.__exit__(exc_type, exc_val, exc_tb)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        try:
+            chunk = self.__wrapped__.__next__()
+            _loop_handler(self._dd_span, chunk, self._streamed_chunks)
+            return chunk
+        except StopIteration:
+            _process_finished_stream(
+                self._dd_integration,
+                self._dd_span,
+                self._args,
+                self._kwargs,
+                self._streamed_chunks,
+                self._is_message_stream,
+            )
+            self._dd_span.finish()
+            self._dd_integration.metric(self._dd_span, "dist", "request.duration", self._dd_span.duration_ns)
+            raise
+        except Exception:
+            self._dd_span.set_exc_info(*sys.exc_info())
+            self._dd_span.finish()
+            self._dd_integration.metric(self._dd_span, "dist", "request.duration", self._dd_span.duration_ns)
+            raise
+
+    def __stream_text__(self):
```

Comment:
Is this a helper method that doesn't exist in `Anthropic.Stream`? If so, we should rename it to `_stream_text`.

Reply:
Yes, `__stream_text__` is not used by `Anthropic.Stream` but is used when calling the `messages.stream` method and is from a helper class of `Anthropic.messages.Stream`. I needed to keep the same function name since we need to override the method. If we don't override it, then we will iterate though the unwrapped stream `Anthropic.messages.Stream` object, and not our `TracedAnthropicStream` object

Reply:
Disregard the above, I renamed it as requested 👍